### PR TITLE
Ensure that side effects of conditional access on an instance of a nullable type are not preserved in generic code

### LINKED ISSUE #66152: Side effects of conditional access on an instance of a nullable type are preserved in generic code
https://sharplab.io/#v2:EYLgtghglgdgNAFxFANgHwLACgDOCBOArgMYIAEAygIzYDe2ZjZADvlAG4QICmZs5xAPaEYCANwMmAAQDMZQe2742AE15SqABjIAVQRQKwA5gAoAlJMb0sTW2SkB2MiYDULoSIRmAdHoNsYUzMJGyYAX2wIrGwpACZ7Kgc6SwSANnsAFjIAWWgYcxTrOyZOfDIADyoyAF4yGG4Ad0oqAH4Teqbqc2CU2x1uPCoTfG4AMwqqHtDijW8NAE4TACIAWjWlqeLGUor42o7mtoOus02t/rxYk3LYs8Yo2xSNdKksi4QqAB4dAD5hsd0FQs0zIRS2swW1xavn0hkC3TEZF60iocyoi3KMP8xgRSJBDyYTyoLzeAwQsW+fx0QMKyMYEPRUKxcKCwTxWwSaIxzICrMRKSiUSAA==
```
struct S1
{
    private int count;
    public override string ToString()
    {
        return (++count).ToString();
    }
}

class Program
{
    static void Main()
    {
        var x1 = new S1?(new S1());
        Test1(ref x1);
        System.Console.WriteLine("---");
        var x2 = new S1?(new S1());
        Test2(x2);
    }
    
    static void Test1<T>(ref T x)
    {
        System.Console.WriteLine(x?.ToString()); 
        System.Console.WriteLine(x.ToString()); 
    }
    
    static void Test2<T>(T x)
    {
        System.Console.WriteLine(x?.ToString()); 
        System.Console.WriteLine(x.ToString()); 
    }
}
```

Observed output:
```
1
1
---
1
2
```

Expected
```
1
1
---
1
1
```

In a non-generic scenario, when receiver is a nullable value type, it is unwrapped and the access is performed on a copy of the underlying non-null value. Therefore, the side effects are never preserved in the original nullable instance. I think Test2 is supposed to make a copy of the receiver when it is a nullable value type at runtime, but, unlike Test1, it doesn't do that.

IL:
```
    .method private hidebysig static 
        void Test1<T> (
            !!T& x
        ) cil managed 
    {
        .param type T
            .custom instance void System.Runtime.CompilerServices.NullableAttribute::.ctor(uint8) = (
                01 00 02 00 00
            )
        // Method begins at RVA 0x2108
        // Code size 71 (0x47)
        .maxstack 2
        .locals init (
            [0] !!T
        )

        IL_0000: ldarg.0
        IL_0001: ldloca.s 0
        IL_0003: initobj !!T
        IL_0009: ldloc.0
        IL_000a: box !!T
        IL_000f: brtrue.s IL_0025

        IL_0011: ldobj !!T
        IL_0016: stloc.0
        IL_0017: ldloca.s 0
        IL_0019: ldloc.0
        IL_001a: box !!T
        IL_001f: brtrue.s IL_0025

        IL_0021: pop
        IL_0022: ldnull
        IL_0023: br.s IL_0030

        IL_0025: constrained. !!T
        IL_002b: callvirt instance string [System.Runtime]System.Object::ToString()

        IL_0030: call void [System.Console]System.Console::WriteLine(string)
        IL_0035: ldarg.0
        IL_0036: constrained. !!T
        IL_003c: callvirt instance string [System.Runtime]System.Object::ToString()
        IL_0041: call void [System.Console]System.Console::WriteLine(string)
        IL_0046: ret
    } // end of method Program::Test1

    .method private hidebysig static 
        void Test2<T> (
            !!T x
        ) cil managed 
    {
        .param type T
            .custom instance void System.Runtime.CompilerServices.NullableAttribute::.ctor(uint8) = (
                01 00 02 00 00
            )
        // Method begins at RVA 0x215b
        // Code size 48 (0x30)
        .maxstack 8

        IL_0000: ldarg.0
        IL_0001: box !!T
        IL_0006: brtrue.s IL_000b

        IL_0008: ldnull
        IL_0009: br.s IL_0018

        IL_000b: ldarga.s x
        IL_000d: constrained. !!T
        IL_0013: callvirt instance string [System.Runtime]System.Object::ToString()

        IL_0018: call void [System.Console]System.Console::WriteLine(string)
        IL_001d: ldarga.s x
        IL_001f: constrained. !!T
        IL_0025: callvirt instance string [System.Runtime]System.Object::ToString()
        IL_002a: call void [System.Console]System.Console::WriteLine(string)
        IL_002f: ret
    } // end of method Program::Test2
```

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #66152.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingOperator.cs [hunk 1]
```diff
@@ -125,6 +125,7 @@ public override BoundNode VisitNullCoalescingOperator(BoundNullCoalescingOperato
                             whenNotNull: notNullAccess,
                             whenNullOpt: whenNullOpt,
                             id: conditionalAccess.Id,
+                            conditionalAccess.ForceCopyOfNullableValueType,
```
Comment (jjonescz) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingOperator.cs:128]:
```suggestion
                            forceCopyOfNullableValueType: conditionalAccess.ForceCopyOfNullableValueType,
```

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_UnaryOperator.cs [hunk 2]
```diff
@@ -276,6 +276,7 @@ public override BoundNode VisitUnaryOperator(BoundUnaryOperator node)
                     whenNotNull: result,
                     whenNullOpt: null,
                     id: conditionalLeft.Id,
+                    conditionalLeft.ForceCopyOfNullableValueType,
```
Comment (jjonescz) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_UnaryOperator.cs:279]:
```suggestion
                    forceCopyOfNullableValueType: conditionalLeft.ForceCopyOfNullableValueType,
```

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review.

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml [hunk 3]
```diff
@@ -1704,6 +1704,7 @@
     matching Id that are integers unique for the containing method body.
     -->
     <Field Name="Id" Type="int"/>
+    <Field Name="ForceCopyOfNullableValueType" Type="bool"/>
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:1707]: :memo: Note to future self: For conditional access we do want a copy when dealing with an unconstrained type, for the nullable value type scenario to behave properly. 
But we also use this bound node in a few other scenarios (string concat and fixed statement) that need different semantics (we shouldn't get any copy).

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 2)

**Review by 333fred — APPROVED** [review]

**Pull request merged**